src/saml2/assertion.py

```diff
@@ -276,162 +277,122 @@ def restriction_from_attribute_spec(attributes):
-                        alist = [k.lower() for k in items]
-                        try:
-                            _only_required = _mod.ONLY_REQUIRED[key]
-                        except (AttributeError, KeyError):
-                            _only_required = False
-                        _ec[key] = (alist, _only_required)
-                    ecs.append(_ec)
-                spec["entity_categories"] = ecs
-            try:
-                restr = spec["attribute_restrictions"]
-            except KeyError:
-                continue
-
-            if restr is None:
-                continue
-
-            _are = {}
-            for key, values in restr.items():
-                if not values:
-                    _are[key.lower()] = None
-                    continue
-
-                _are[key.lower()] = [re.compile(value) for value in values]
-            spec["attribute_restrictions"] = _are
-        logger.debug("policy restrictions: %s", self._restrictions)
+    def setup_restrictions(self, restrictions=None):
+        if restrictions is None:
+            return None
 
-        return self._restrictions
+        restrictions = copy.deepcopy(restrictions)
+        restrictions = compile(restrictions)
+        return restrictions
 
-    def get(self, attribute, sp_entity_id, default=None, post_func=None,
-            **kwargs):
+    def get(self, attribute, sp_entity_id, default=None):
         """
 
         :param attribute:
         :param sp_entity_id:
         :param default:
-        :param post_func:
         :return:
         """
         if not self._restrictions:
             return default
 
-        try:
-            try:
-                val = self._restrictions[sp_entity_id][attribute]
-            except KeyError:
-                try:
-                    val = self._restrictions["default"][attribute]
-                except KeyError:
-                    val = None
-        except KeyError:
-            val = None
-
-        if val is None:
-            return default
-        elif post_func:
-            return post_func(val, sp_entity_id=sp_entity_id, **kwargs)
-        else:
-            return val
+        ra_info = (
+            self.metadata_store.registration_info(sp_entity_id) or {}
+            if self.metadata_store is not None
+            else {}
+        )
+        ra_entity_id = ra_info.get("registration_authority")
+
+        sp_restrictions = self._restrictions.get(sp_entity_id)
+        ra_restrictions = self._restrictions.get(ra_entity_id)
+        default_restrictions = (
+            self._restrictions.get("default")
+            or self._restrictions.get("")
+        )
+        restrictions = (
+            sp_restrictions
+            if sp_restrictions is not None
+            else ra_restrictions
+            if ra_restrictions is not None
+            else default_restrictions
+            if default_restrictions is not None
+            else {}
```

Comment:
Use the `sp_restrictions`, else the `ra_restrictions`, else the `default_restrictions`.